qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,141 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Cancel the redundant (self-adjoint) gates through commutation relations."""
+
+from collections import defaultdict
+
+import numpy as np
+
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.circuit.commutation_checker import CommutationChecker
+from qiskit.circuit.commutation_library import standard_gates_commutations
+from qiskit.circuit.controlflow import CONTROL_FLOW_OP_NAMES
+from qiskit.circuit.instruction import Instruction
+from qiskit.circuit.library.generalized_gates.pauli import PauliGate
+from qiskit.circuit.library.standard_gates.p import PhaseGate
+from qiskit.circuit.library.standard_gates.rx import RXGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.library.standard_gates.u1 import U1Gate
+from qiskit.circuit.quantumregister import QuantumRegister, Qubit
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.dagcircuit import DAGCircuit, DAGInNode, DAGNode, DAGOutNode
+from qiskit.quantum_info import Pauli
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.passes.optimization.commutation_analysis import \
+    CommutationAnalysis
+from qiskit.transpiler.passmanager import PassManager
+
+commutator = CommutationChecker(standard_gates_commutations)
+
+
+class LightCone(TransformationPass):
+    """Remove the gates that do not affect the outcome of a measurement on a circuit.
+
+    Pass for computing the light-cone of an observable or measurement. The Pass can
+    handle either a Pauli operator one would like to measure or a measurement on a set
+    of qubits.
+    """
+
+    def __init__(self, observable: Pauli | None = None):
+        """
+        LightCone initializer.
+
+        Args:
+            observable: If None the lightcone will be computed for the set
+            of measurements in the circuit.
+            If a Pauli operator is specified, the lightcone will correspond to
+            the reduced circuit with the
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
```

Comment:
We can do this more efficiently if we only consider final measurements, see how it's done in the `RemoveFinalMeasurements` pass 🙂

Reply:
I thought about it. `RemoveFinalMeasurements` has a function `_calc_final_ops` that would be ideal to reuse. Can I change the class `RemoveFinalMeasurements` to have that function outside of the class so that I can call it from this other class or would it be better to copy-paste the function into `LightCone`?

Reply:
We could just make the function public, it seems generally useful. In this case we should give it a more precise name, e.g. something like `final_measurement_nodes`, and add a short test for it.